File: go/internal/obs/metrics.go
// Package obs is a tiny dependency-free metrics registry exposing Prometheus text format.
package obs

import (
	"fmt"
	"sort"
	"sync"
)

// Metrics is a concurrency-safe counter/gauge registry.
type Metrics struct {
	mu       sync.Mutex
	counters map[string]float64
}

// New builds an empty registry.
func New() *Metrics { return &Metrics{counters: map[string]float64{}} }

// Inc adds delta to a counter.
func (m *Metrics) Inc(name string, delta float64) {
	m.mu.Lock()
	defer m.mu.Unlock()
	m.counters[name] += delta
}

// Get returns a counter's value.
func (m *Metrics) Get(name string) float64 {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.counters[name]
}

// Expose renders the registry in Prometheus text format.
func (m *Metrics) Expose() string {
	m.mu.Lock()
	defer m.mu.Unlock()
	names := make([]string, 0, len(m.counters))
	for n := range m.counters {
		names = append(names, n)
	}
	sort.Strings(names)
	var b []byte
	for _, n := range names {
		b = append(b, fmt.Sprintf("# TYPE %s counter\n%s %g\n", n, n, m.counters[n])...)
	}
	return string(b)
}

File: csharp/src/GuardianForge.Core/Metrics.cs
namespace GuardianForge.Core;

/// <summary>A tiny concurrency-safe counter registry exposing Prometheus text format.</summary>
public sealed class Metrics
{
    private readonly object _lock = new();
    private readonly Dictionary<string, double> _counters = new();

    public void Inc(string name, double delta)
    {
        lock (_lock) _counters[name] = _counters.GetValueOrDefault(name) + delta;
    }

    public double Get(string name)
    {
        lock (_lock) return _counters.GetValueOrDefault(name);
    }

    public string Expose()
    {
        lock (_lock)
        {
            var sb = new System.Text.StringBuilder();
            foreach (var name in _counters.Keys.OrderBy(k => k, StringComparer.Ordinal))
                sb.Append($"# TYPE {name} counter\n{name} {_counters[name]}\n");
            return sb.ToString();
        }
    }
}
